visionmetrics/detection/confusion_matrix.py

```diff
@@ -0,0 +1,171 @@
+        targets = [[[0, 10, 20, 50, 100], [1, 30, 40, 80, 120]], [[1, 20, 30, 60, 90]]]
+
+        Returns
+        -------
+        confusion_matrix : dict
+            Dictionary containing the counts of TP, FP, and FN, as well as details for FP reasons
+
+            TP: When a predicted bounding box has IoU greater than the threshold with a ground truth box of the same class
+            FN: When a ground truth bounding box has no corresponding predicted bounding box
+            FP:
+                1. FP_due_to_wrong_class: IOU >= threshold but predicted class is different from the ground truth class.
+                2. FP_due_to_low_iou: Predicted bbox IOU < threshold (including no overlap)
+                3. fp_due_to_extra_pred_boxes: Excess predicted bboxes when all ground truth boxes have been matched
+
+        """
+
+    def __init__(self, iou_threshold=0.5, **kwargs):
+        super().__init__(**kwargs)
+        self.iou_threshold = iou_threshold
+        self.add_state("tp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fn", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_wrong_class", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_low_iou", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_extra_pred_boxes", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, predictions, targets):
+        if len(predictions) != len(targets):
+            raise ValueError("Number of predictions and targets should be the same.")
+        if not isinstance(predictions[0][0], list):
+            raise ValueError(f"Expected predictions to be a list of lists, got a list of {type(predictions[0][0])}")
+
+        self._update_confusion_matrix(predictions, targets, self.iou_threshold)
+
+    def compute(self):
+        return {
+            'TP': self.tp.item(),
+            'FP': self.fp.item(),
+            'FN': self.fn.item(),
+            'FP_due_to_wrong_class': self.fp_due_to_wrong_class.item(),
+            'FP_due_to_low_iou': self.fp_due_to_low_iou.item(),
+            'FP_due_to_extra_pred_boxes': self.fp_due_to_extra_pred_boxes.item()
+        }
+
+    def _update_confusion_matrix(self, predictions, targets, iou_threshold):
+        for preds, gts in zip(predictions, targets):
+
+            # Check empty predictions and targets
+            if self._is_empty(preds):
+                self.fn += len(gts)
+                continue
+
+            if self._is_empty(gts):
+                self.fp += len(preds)
+                self.fp_due_to_extra_pred_boxes += len(preds)
+                continue
+
+            # Count per image
+            gt_boxes_used = [False] * len(gts)  # To keep track of all matched GT boxes
+            preds = sorted(preds, key=lambda x: x[1], reverse=True)
+            for pred in preds:
```

Comment:
You are evaluating each prediction against each gt, which is kind of O(n²). I wonder if you could use something like this: https://msazure.visualstudio.com/VisionIntelligence/_git/RaaS.ModelTraining.Shared?path=/evaluation/detection_evaluation.py&version=GBmaster&line=395&lineEnd=396&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents to vectorize this op.

Reply:
It's hard to vectorize since the inner loop is variable with multiple `continue` and `break` conditions. 
I can revisit if O(n2) becomes a bottleneck.